### ISSUE TITLE
Transformers Trainer.train() gets stuck when executed on notebook server

### ISSUE THREAD

**Issue by pthalasta** [kubeflow/kubeflow#5638]
I'm trying to train a transformer based on BertForSequenceClassification model in a notebook spun up using kubeflow. 
When i try to start the training of the model there is no output or error shown by the function even after 30mins of runtime. GPU usage doesn't change as well

Script/sample notebook executed: https://github.com/huggingface/notebooks/blob/master/examples/text_classification.ipynb
transformer version used in the notebook

```
transformers version: 4.3.2
Platform: Linux-4.15.0-76-generic-x86_64-with-debian-buster-sid
Python version: 3.7.10
PyTorch version (GPU?): 1.7.1 (True)
```

I also tried executing the same notebook on a docker container and it totally works fine. How do i resolve the issue?

**Comment by DavidSpek** [comment]
Which notebook did you use? Are you able to run other notebooks that use the GPU successfully?

**Comment by pthalasta** [comment]
@DavidSpek i'm not sure what you meant by which notebook? Do you mean the version?
The issue exists with all the notebooks that use transformers in order to train a model. Pytorch can see the GPU devices as the output of `torch.cuda.device_count()` was valid.

**Comment by DavidSpek** [comment]
@pthalasta Is this one of the notebook images provided with Kubeflow? Or is it a custom one? If so, can you share the Dockerfile so I can take a look at it?

**Comment by pthalasta** [comment]
@DavidSpek i use RAPIDS image and override the entry point as specified by kubeflow for custom images.
Docker file: https://github.com/rapidsai/clx/blob/branch-0.18/Dockerfile

**Comment by DavidSpek** [comment]
I have no experience with rapids, but I will try to look at it tomorrow. However, it seems like you are installing a nightly build which I assume is probably not stable and could cause unexpected issues.

**Comment by pthalasta** [comment]
@DavidSpek any update on this?

**Comment by pthalasta** [comment]
@DavidSpek I also tried with non-nightly image. 

Docker pull for the image: `docker pull rapidsai/rapidsai-clx:0.18-cuda11.0-runtime-ubuntu20.04-py3.8`
I still see the same issue with this as well

**Comment by DavidSpek** [comment]
@pthalasta Sorry I haven't been able to look at it much yet. Can you please share the actual docker file you use, along with a startup script if you use that.

**Comment by pthalasta** [comment]
@DavidSpek here are the details
Dockerfile for base image used: https://github.com/rapidsai/docker/blob/branch-0.19/generated-dockerfiles/rapidsai-clx_ubuntu20.04-devel.Dockerfile

Dockerfile for the image that is used in kubeflow:

```
FROM rapidsai/rapidsai-clx:0.18-cuda11.0-runtime-ubuntu20.04-py3.8


COPY entrypoint.sh /opt/docker/bin/entrypoint
RUN chmod +x /opt/docker/bin/entrypoint
ENTRYPOINT [ "/usr/bin/tini", "--", "/opt/docker/bin/entrypoint" ]
```

Entrypoint script used:

```
#!/bin/bash

set -e
source activate rapids
exec "$@"
```

**Comment by pthalasta** [comment]
@DavidSpek any update on this issue?

**Comment by DavidSpek** [comment]
@pthalasta I'm very sorry that I haven't gotten around to this yet. All my time and effort is going into an unproductive back and forth with the @kubeflow/wg-notebooks-leads about the work I have done that would help solve this very issue. However, I will build the image you are using and try to replicate the problem you are facing tomorrow.

**Comment by DavidSpek** [comment]
@pthalasta Again I am sorry to have to say this, but my day has gotten so busy again that I haven't been able to look at it yet. However, I will get to this tomorrow or Saturday.

**Comment by pthalasta** [comment]
@DavidSpek sure, no worries! Thanks for the update.

**Comment by DavidSpek** [comment]
@pthalasta First thing, the `CMD` from your Dockerfile above contains an error. It should be: 
```
CMD ["sh","-c","jupyter-lab --notebook-dir=/home/jovyan --ip=0.0.0.0 --no-browser --allow-root --port=8888 --NotebookApp.token='' --NotebookApp.password='' --NotebookApp.allow_origin='*' --NotebookApp.base_url=${NB_PREFIX}"]
```

I built the container and am running the Notebook you linked in the first post without a problem on Kubeflow 1.2 using Kubernetes 1.20.
```
Mon Mar 15 10:48:04 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.39       Driver Version: 460.39       CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GeForce RTX 3090    Off  | 00000000:42:00.0 Off |                  N/A |
| 49%   56C    P2   136W / 350W |   2663MiB / 24268MiB |     19%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

To narrow down what is causes the problem I think it might be best to build your own container with the tools you need, this way you know everything that is installed. I would recommend you start from the images listed in https://github.com/kubeflow/kubeflow/pull/5626. You can use https://github.com/kubeflow/kubeflow/pull/5631 as an example how to add packages. Those Dockerfile uses pip, but you can also use conda if you prefer that (in my experience conda creates much larger images, which is why pip is being used).

**Comment by pthalasta** [comment]
@DavidSpek Thanks for the update! I fixed the error in cmd that you mentioned. However, the when i try to execute the notebook as per https://github.com/huggingface/notebooks/blob/master/examples/text_classification.ipynb , it gets stuck when `Trainer.train()` is invoked. 

The docker image that i have built is able to see the GPUs without any issue

**Comment by DavidSpek** [comment]
@pthalasta Are you able to load a dataset into the GPU memory? It will take a bit of debugging work to figure out what is happening here.

**Comment by pthalasta** [comment]
@DavidSpek yes, i can load the dataset into GPU memory.

**Comment by DavidSpek** [comment]
Are you able to train other models on the GPU? Can you get more debug outputs (In JupyterLab, under `View` there is an option `Show Log Conolse`, set Log lLevel to Debug and run the code)? Where did you deploy Kubeflow? What version of Kubernetes?

**Comment by pthalasta** [comment]
@DavidSpek thanks for suggestion! I went ahead and tried to see the logs while executing the code for model training. I do not see any errors being logged. Also, i did a quick check that i can train other models and used this training script provided by tensorflow as an example https://www.tensorflow.org/tutorials/quickstart/advanced and it works fine and i can see the variation in GPU utilization metric with time during the training.

I'm currently using kubeflow build version `v1beta1` and kubernters version is `1.18`

**Comment by DavidSpek** [comment]
@pthalasta I might be mistaken, but I thought `rapidsai` was using PyTorch and not TensorFlow. If that is the case, can you confirm PyTorch works with your GPU in that image?

The build version you are seeing is likely the dashboard version, the kubeflow version would be 1.1 or 1.2 most likely. Are you running this in GKE, EKS, AKS, OpenShift, or something else?

**Comment by pthalasta** [comment]
@DavidSpek apologies as i misunderstood the question. Yes, pytorch can use GPUs in the image created without any issues and i tried it using the code provided here (sample code taken from https://pytorch.org/tutorials/beginner/pytorch_with_examples.html)

```
import torch
import math


dtype = torch.float
device = torch.device("cuda:0") # Uncomment this to run on GPU

# Create random input and output data
x = torch.linspace(-math.pi, math.pi, 2000, device=device, dtype=dtype)
y = torch.sin(x)

# Randomly initialize weights
a = torch.randn((), device=device, dtype=dtype)
b = torch.randn((), device=device, dtype=dtype)
c = torch.randn((), device=device, dtype=dtype)
d = torch.randn((), device=device, dtype=dtype)

learning_rate = 1e-6
for t in range(2000):
    # Forward pass: compute predicted y
    y_pred = a + b * x + c * x ** 2 + d * x ** 3

    # Compute and print loss
    loss = (y_pred - y).pow(2).sum().item()
    if t % 100 == 99:
        print(t, loss)

    # Backprop to compute gradients of a, b, c, d with respect to loss
    grad_y_pred = 2.0 * (y_pred - y)
    grad_a = grad_y_pred.sum()
    grad_b = (grad_y_pred * x).sum()
    grad_c = (grad_y_pred * x ** 2).sum()
    grad_d = (grad_y_pred * x ** 3).sum()

    # Update weights using gradient descent
    a -= learning_rate * grad_a
    b -= learning_rate * grad_b
    c -= learning_rate * grad_c
    d -= learning_rate * grad_d


print(f'Result: y = {a.item()} + {b.item()} x + {c.item()} x^2 + {d.item()} x^3')
```

Regarding the version, kubeflow is 1.2 and its running on one of on-prem node with NVIDIA GPU(Tesla V100).

**Comment by DavidSpek** [comment]
@pthalasta Thanks for the information. Would you be able to try `public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-pytorch-cuda-full:master-28af7716`? You will need to manually install `datasets` and `transformers` using the first cell (remove the `#` and the `,`), then restart the kernel and see if it runs. If it doesn't, can you try your code above or the code below to test if that works?
```
from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.optim.lr_scheduler import StepLR

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout(0.25)
        self.dropout2 = nn.Dropout(0.5)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output

def train(model, device, train_loader, optimizer, epoch, log_interval, dry_run):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.item()))
            if dry_run:
                break

def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))

def main(train_batch_size, test_batch_size, epochs, lr, gamma, no_cuda, dry_run, seed, log_interval, save_model):
    use_cuda = not no_cuda #and torch.cuda.is_available()

    torch.manual_seed(seed)

    device = torch.device("cuda" if use_cuda else "cpu")

    train_kwargs = {'batch_size': train_batch_size}
    test_kwargs = {'batch_size': test_batch_size}
    if use_cuda:
        cuda_kwargs = {'num_workers': 1,
                       'pin_memory': True,
                       'shuffle': True}
        train_kwargs.update(cuda_kwargs)
        test_kwargs.update(cuda_kwargs)

    transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
    dataset1 = datasets.MNIST('./data', train=True, download=True,
                       transform=transform)
    dataset2 = datasets.MNIST('./data', train=False,
                       transform=transform)
    train_loader = torch.utils.data.DataLoader(dataset1,**train_kwargs)
    test_loader = torch.utils.data.DataLoader(dataset2, **test_kwargs)

    model = Net().to(device)
    optimizer = optim.Adadelta(model.parameters(), lr=lr)

    scheduler = StepLR(optimizer, step_size=1, gamma=gamma)
    for epoch in range(1, epochs + 1):
        train(model, device, train_loader, optimizer, epoch, log_interval, dry_run)
        test(model, device, test_loader)
        scheduler.step()

    if save_model:
        torch.save(model.state_dict(), "mnist_cnn.pt")

main(train_batch_size=64,
     test_batch_size=1000,
     epochs=2, lr=1.0,
     gamma=0.7,
     no_cuda=False,
     dry_run=False,
     seed=1,
     log_interval=10,
     save_model=False)
```

**Comment by pthalasta** [comment]
@DavidSpek i executed the code provided above by you and it works fine without any issues on the image that you mentioned. Once i confirmed that was working fine, i went ahead and tried executing the notebook as per https://github.com/huggingface/notebooks/blob/master/examples/text_classification.ipynb and it gets stuck on `trainer.train()` even on the image `public.ecr.aws/j1r0q0g6/notebooks/notebook-servers/jupyter-pytorch-cuda-full:master-28af7716` and i do not see any errors in the console.

**Comment by pthalasta** [comment]
@DavidSpek i'm not the administrator, @jolorunyomi can provide more information on this.

Yes, i created the issue with package repo thinking it might be an issue there. But, i was able to execute the notebook without any issue on the docker container and hence created this issue.

**Comment by DavidSpek** [comment]
@pthalasta Given that I am able to execute the notebook without any issue with the same docker container in Kubernetes and within Kubeflow I don't think it is a Kubeflow issue. As I can't replicate the problem I also fear there is not much that I can do to help other than think of what might be causing the problem on that node specifically.

**Comment by pthalasta** [comment]
@DavidSpek thanks for the information! I found out from the administrator that we are running kubeflow version `1.02`. Can that be a reason for the issue i'm having?

**Comment by DavidSpek** [comment]
@pthalasta I highly doubt it. I think it is more likely related to the hardware, drivers, OS or container runtime and the transformers package. Kubeflow doesn't do anything with the notebook containers that you run other than schedule them and setup the routing so you can access it.

**Comment by pthalasta** [comment]
@DavidSpek i did a little more testing for this. So, the trainer.train() works when the notebook is created with 1 GPU but not when there are multiple GPUs available. Just to confirm, when you tested in your environment, was it 1 GPU instance or multi-GPU?

**Comment by DavidSpek** [comment]
It was a single GPU instance. I am testing this on my home cluster which only contains a single GPU. However, I still believe this is most likely an issue with the transformers package and multi-GPU in this case.
